fireworks/scripts/lpad_run.py

```diff
@@ -691,6 +692,115 @@ def maintain(args):
     lp.maintain(args.infinite, args.maintain_interval)
 
 
+def orphaned(args):
+    # get_fws
+    lp = get_lp(args)
+    if sum([bool(x) for x in [args.fw_id, args.name, args.state, args.query]]) > 1:
+        raise ValueError('Please specify exactly one of (fw_id, name, state, query)')
+    if sum([bool(x) for x in [args.fw_id, args.name, args.state, args.query]]) == 0:
+        args.query = '{}'
+        args.display_format = args.display_format if args.display_format else 'ids'
+    if sum([bool(x) for x in [args.fw_id, args.name, args.qid]]) > 1:
+        raise ValueError('Please specify exactly one of (fw_id, name, qid)')
+    else:
+        args.display_format = args.display_format if args.display_format else 'more'
+
+    if args.fw_id:
+        query = {'fw_id': {"$in": args.fw_id}}
+    elif args.name and not args.launches_mode:
+        query = {'name': args.name}
+    elif args.state:
+        query = {'state': args.state}
+    elif args.query:
+        query = ast.literal_eval(args.query)
+    else:
+        query = None
+
+    if args.sort:
+        sort = [(args.sort, ASCENDING)]
+    elif args.rsort:
+        sort = [(args.rsort, DESCENDING)]
+    else:
+        sort = None
+
+    if args.qid:
+        fw_ids = lp.get_fw_ids_from_reservation_id(args.qid)
+        if query:
+            query['fw_id'] = {"$in": fw_ids}
+            fw_ids = lp.get_fw_ids(query, sort, args.max, launches_mode=args.launches_mode)
+    else:
+        fw_ids = lp.get_fw_ids(query, sort, args.max,
+                               launches_mode=args.launches_mode)
+
+    # get_wfs
+    orphaned_fw_ids = []
+    for fw_id in fw_ids:
+        query = {'nodes': fw_id}
+        wf_ids = lp.get_wf_ids(query)
+        if len(wf_ids) == 0:
+            orphaned_fw_ids.append(fw_id)
```

Comment:
This seems like it would be really slow for a large FireWorks database, e.g., if the potentially empty query returned 1 million results.

If the query returns more than 100 results or so, it might be quicker to simply make a list of all the ```fw_ids``` present in the database overall (e.g., do an empty query on the ```workflows database```, and get back the nodes parameter). Then you have two set of ```fw_ids``` - the set matching the query, and the set of all fw_ids in the workflows collection. You can use Python's set difference to get the ones that are in the query but not in the workflows collection.

Let me know what you think about this

Reply:
Very true, some proper joint aggregation query on the fireworks and workflows collections in the db might be  even better, but I won't do that for now.